tidy3d/plugins/adjoint/components/medium.py

```diff
@@ -0,0 +1,288 @@
+    def _single_frequency(cls, val):
+        return val
+
+    @pd.validator("eps_dataset", always=True)
+    def _eps_inf_greater_no_less_than_one_sigma_positive(cls, val):
+        return val
+
+    def eps_dataset_freq(self, frequency: float) -> PermittivityDataset:
+        as_custom_medium = self.to_medium()
+        return as_custom_medium.eps_dataset_freq(frequency=frequency)
+
+    def to_medium(self) -> CustomMedium:
+        """Convert :class:`.JaxMedium` instance to :class:`.Medium`"""
+        self_dict = self.dict(exclude={"type"}).copy()
+        eps_field_components = {}
+        for dim in "xyz":
+            field_name = f"eps_{dim}{dim}"
+            data_array = self_dict["eps_dataset"][field_name]
+            values = data_array["values"]
+            coords = data_array["coords"]
+            scalar_field = ScalarFieldDataArray(values, coords=coords)
+            eps_field_components[field_name] = scalar_field
+        eps_dataset = PermittivityDataset(**eps_field_components)
+        self_dict["eps_dataset"] = eps_dataset
+        return CustomMedium.parse_obj(self_dict)
+
+    @classmethod
+    def from_tidy3d(cls, tidy3d_obj: CustomMedium) -> JaxCustomMedium:
+        """Convert :class:`.Tidy3dBaseModel` instance to :class:`.JaxObject`."""
+        obj_dict = tidy3d_obj.dict(exclude={"type", "eps_dataset"}).copy()
+        eps_dataset = tidy3d_obj.eps_dataset
+        field_components = {}
+        for dim in "xyz":
+            field_name = f"eps_{dim}{dim}"
+            data_array = eps_dataset.field_components[field_name]
+            values = data_array.values.tolist()
+            coords = {key: np.array(val).tolist() for key, val in data_array.coords.items()}
+            field_components[field_name] = JaxDataArray(values=values, coords=coords)
+        eps_dataset = JaxPermittivityDataset(**field_components)
+        obj_dict["eps_dataset"] = eps_dataset
+        return cls.parse_obj(obj_dict)
+
+    # pylint:disable=too-many-locals, unused-argument
+    def store_vjp(
+        self,
+        grad_data_fwd: FieldData,
+        grad_data_adj: FieldData,
+        sim_bounds: Bound,
+        wvl_mat: float,
+    ) -> JaxMedium:
+        """Returns the gradient of the medium parameters given forward and adjoint field data."""
+        vjp_field_components = {}
+        for dim in "xyz":
+
+            eps_field_name = f"eps_{dim}{dim}"
+            field_name = f"E{dim}"
+
+            orig_data_array = self.eps_dataset.field_components[eps_field_name]
+            coords = orig_data_array.coords
+            interp_coords = {dim_pt: coords[dim_pt] for dim_pt in "xyz" if len(coords[dim_pt]) > 1}
+            isel_coords = {dim_pt: 0 for dim_pt in "xyz" if len(coords[dim_pt]) <= 1}
+            e_fwd = grad_data_fwd.field_components[field_name]
+            e_adj = grad_data_adj.field_components[field_name]
+            e_dotted = (e_fwd * e_adj).isel(f=0, **isel_coords).interp(**interp_coords)
```

Comment:
As discussed, this seems to work only in a 2D simulation. In a 3D simulation, I think `.isel` should be replaced with a integration along this axis. But we can leave it to the next release and add a warning here.

Reply:
Yea, I think more specifically, this is only totally accurate in the limit that the influence of the eps_dataset in the CustomMedium approaches size of (0,0,0). So for very small pixels along each of the dimensions (including 2D) this works, but in the future we will want to integrate over the "influence" of the pixel. This integration volume probably will depend on whether "nearest" or "inerpolate" is used in the CustomMedium. For nearest, just integrate over the current "pixel", but for interpolate, integrate over the pixel and its neighbors. Does this seem about right? For now, it's working so well in the case study that I might add a warning and come back to it later.

Reply:
I just added a warning to address this when `JaxCustomMedium` is constructed:
```py
    @pd.root_validator(always=True, pre=True)
    def _warn_about_pixel_size(cls, values):
        """Warning about using :class:`.CustomMedium`."""
        log.warning(
            "'JaxCustomMedium' currently assumes the size of each pixel in 'eps_dataset' is "
            "small compared to the spatial features in the electric field distribution. "
            "Therefore, it is well-suited for inverse design optimization with small pixel sizes, "
            "but for modeling larger pixels, it is better to use several 'JaxBox' objects instead."
        )
        return values
```

Reply:
>  but for interpolate, integrate over the pixel and its neighbors. Does this seem about right?

Yes, and with some weight. This part can probably be handled by Jax if VJP for xarray's interp is defined?

> I just added a warning to address this when JaxCustomMedium is constructed

Looks good. Shall we add a warning specifically for the case where only one entry is supplied along any axis unless the thickness of geometry is 1 pixel along that axis?

Reply:
> Yes, and with some weight. This part can probably be handled by Jax if VJP for xarray's interp is defined?

In principle, this would be a great way to do it, however, I had to make my own `XR.DataArray` clone called `JaxDataArray` and I'm not really sure how to implement the `.interp()` method or its vjp, especially with Jax tracers. This might be a bit complicated and maybe we should revisit later.

> Shall we add a warning specifically for the case where only one entry is supplied along any axis unless the thickness of geometry is 1 pixel along that axis?

I think we should always warn upon creation of a CustomMedium. Here's what's going on.
Line 268 is telling Jax how to evaluate the VJP of that pixel. The approximation here is that the VJP of that pixel should be proportional to the E_fwd * E_adj evaluated at that pixel coordinates. The proper VJP is actually the weighted average of this quantity over the whole pixel volume. As long as the field does not vary much over this pixel, they should be equivalent. So the case where one entry is supplied is fine if the geometry has no thickness along that dimension or the simulation is translationally invariant along that dimension. However, we still have issues if the pixel **size** is too large along the other dimensions. Because the E_fwd and E_adj will vary and we can't approximate the integral with a single value. So therefore, I think it's basically important to warn the user that (for now) they need to ensure they don't expect much field variation within the pixel as a more general warning.